ukat/mapping/diffusion.py

```diff
@@ -0,0 +1,414 @@
+            Magn Reson Mater Phy 2020;33:177–195
+            doi: 10.1007/s10334-019-00790-y.
+            If False, all b-values supplied will be used to fit ADC.
+        """
+        ukrin_b_test = np.array([0, 100, 200, 800])
+        # Sanity checks
+        assert (pixel_array.shape[-1]
+                == len(bvals)), 'Number of bvals does not match number of ' \
+                                'gradients in pixel_array'
+
+        if ukrin_b:
+            self.b_mask = np.isin(bvals, ukrin_b_test)
+        else:
+            self.b_mask = np.full(len(bvals), True, dtype=bool)
+
+        self.pixel_array = pixel_array[..., self.b_mask]
+        self.shape = pixel_array.shape[:-1]
+        self.n_vox = np.prod(self.shape)
+        self.bvals = bvals[self.b_mask]
+        self.n_grad = len(self.bvals)
+        self.u_bvals = unique_bvals_tolerance(self.bvals, 1)
+        self.n_bvals = len(self.u_bvals)
+        self.affine = affine
+        # Generate a mask if there isn't one specified
+        if mask is None:
+            self.mask = np.ones(self.shape, dtype=bool)
+        else:
+            self.mask = mask
+            # Don't process any nan values
+        self.mask[np.isnan(np.sum(pixel_array, axis=-1))] = False
+        self.pixel_array = np.nan_to_num(self.pixel_array)
+
+        self.pixel_array_mean = self.__mean_over_directions__()
+
+        self.adc, self.adc_err = self.__fit__()
+
+    def __mean_over_directions__(self):
+        """
+        Calculates the mean signal across different directions at each unique
+        b-value e.g. if `pixel_array` contains six volumes acquired with
+        b=600 s/mm^2 in different directions, these six volumes will be
+        averaged together.
+
+        Returns
+        -------
+        pixel_array_mean : np.ndarray
+            The average of the `pixel_array` across bvecs
+        """
+        pixel_array_mean = np.zeros((*self.shape, self.n_bvals))
+        for ind, bval in enumerate(self.u_bvals):
+            pixel_array_mean[..., ind]\
+                = np.mean(self.pixel_array[..., self.bvals == bval], axis=-1)
+        return pixel_array_mean
+
+    def __fit__(self):
+        # Initialise maps
+        adc_map = np.zeros(self.n_vox)
+        adc_err = np.zeros(self.n_vox)
+
+        mask = self.mask.flatten()
+        signal = self.pixel_array_mean.reshape(-1, self.n_bvals)
+        idx = np.argwhere(mask).squeeze()
+        with tqdm(total=idx.size) as progress:
+            for ind in idx:
+                sig = signal[ind, :]
+                adc_map[ind], adc_err[ind] = \
+                    self.__fit_signal__(sig, self.u_bvals)
+                progress.update(1)
+        adc_map[adc_map < 0] = 0
+        adc_err[adc_map < 0] = 0
+
+        # Reshape results into raw data shape
+        adc_map = adc_map.reshape(self.shape)
+        adc_err = adc_err.reshape(self.shape)
+
+        return adc_map, adc_err
+
+    @staticmethod
+    def __fit_signal__(sig, bvals):
```

Comment:
Out of curiosity, does DIPY have a function that calculates ADC?

Reply:
No, no monoexponential fitting model. Dipy is more focused in DTI with a bit of IVIM.

Reply:
I see, cheers!